src/storage/src/compacter.rs

```diff
@@ -0,0 +1,691 @@
+        read_segment(&self.path)
+    }
+
+    /// Physically concatenate all batches together into a single batch
+    /// and consolidate updates up to the frontier if provided.
+    fn compact(
+        batches: &[Batch],
+        trace_path: &Path,
+        compaction_frontier: Option<Timestamp>,
+    ) -> Result<Self, anyhow::Error> {
+        let mut messages = vec![];
+
+        for batch in batches {
+            messages.append(&mut read_segment(&batch.path)?);
+        }
+
+        Batch::create_from_messages(messages, trace_path, compaction_frontier)
+    }
+}
+
+/// A Trace is an on-disk representation of data meant to mimic a differential Trace.
+///
+/// A Trace listens to the `wal_path` and checks for WAL segments that it can consolidate
+/// into Batches (stored in the `trace_path`). Once it exceeds a certain number of Batches
+/// it tries to physically and logically compact them into a single batch that is
+/// compacted up to the `compaction` frontier.
+#[derive(Debug)]
+pub struct Trace {
+    trace_path: PathBuf,
+    wal_path: PathBuf,
+    batches: Vec<Batch>,
+    compaction: Option<Timestamp>,
+}
+
+impl Trace {
+    fn create(id: GlobalId, trace_path: PathBuf, wal_path: PathBuf) -> Result<Self, anyhow::Error> {
+        let _ = fs::read_dir(&wal_path).with_context(|| {
+            format!(
+                "trying to ensure wal directory {} exists for trace of relation {}",
+                id,
+                wal_path.display()
+            )
+        })?;
+
+        // Create a new directory to store the trace
+        fs::create_dir(&trace_path).with_context(|| {
+            format!("trying to create trace directory: {}", trace_path.display())
+        })?;
+
+        Ok(Self {
+            trace_path,
+            wal_path,
+            batches: Vec::new(),
+            compaction: None,
+        })
+    }
+
+    /// Remove all on-disk data for this trace.
+    ///
+    /// Importantly, we also delete the WAL directory here (the WAL writer
+    /// only gets to add new files and can't do anything else)
+    fn destroy(self) -> Result<(), anyhow::Error> {
+        fs::remove_dir_all(&self.trace_path).with_context(|| {
+            format!(
+                "failed to remove trace directory {}",
+                self.trace_path.display()
+            )
+        })?;
+        fs::remove_dir_all(&self.wal_path).with_context(|| {
+            format!("failed to remove wal directory {}", self.wal_path.display())
+        })?;
+
+        Ok(())
+    }
+
+    /// Checks if there are finished WAL segments and if so, forms them into batches.
+    fn consume_wal(&mut self) -> Result<(), anyhow::Error> {
+        let finished_segments = self.find_finished_wal_segments()?;
+
+        for segment in finished_segments {
+            let batch = Batch::create(&segment, &self.trace_path)?;
+            self.batches.push(batch);
+            fs::remove_file(&segment).with_context(|| {
```

Comment:
I can believe it is fine, but here is where I'd love to see comments like `// batch creation ensures that it is durably stored; this sequencing ensures that we only remove the wal segment after this happens`.

Reply:
I think this is also worryingly non-atomic. If this crashes after minting the batch but before deleting the wal segment, what happens? Could be that as the segment is finalized determinism says we'd produce the same batch and the error handling there handles it, but talking that out here seems good (and flagging it as a moment of concern seems good too).

More generally, idk what the invariants should be for the on-disk data. It could be that they are great! (like: nothing *needs* to be deleted anywhere; just be smart about what you read) and if that is true super and if not uh oh but flag it.

Reply:
yeah yeah this code doesn't check the invariants but i think there is enough specification that you don't need to delete anything:

* log segments should span contiguous intervals of `[lower, upper)`
* the data in log segments needs to be in time order (no update at t should come after t was closed)
* batches should span contiguous intervals of `[lower, upper)`
* we only need to keep exactly one batch or log segment for each timestamp interval
...

I kinda ran out of steam to write the "check and repair all invariants on restart logic" but 100% this is my next step

Reply:
wrote the logic to check and verify these invariants